.github/workflows/root-ci-config/build_root.py

```diff
@@ -379,24 +407,127 @@ def create_binaries(buildtype):
 
 
 @github_log_group("Rebase")
-def rebase(base_ref, head_ref) -> None:
-    head_ref_src, _, head_ref_dst = head_ref.partition(":")
-    head_ref_dst = head_ref_dst or "__tmp"
+def rebase(directory: str, repository:str, base_ref: str, head_ref: str, head_sha: str) -> None:
     # rebase fails unless user.email and user.name is set
     result = subprocess_with_log(f"""
-        cd '{WORKDIR}/src'
+        cd '{WORKDIR}/{directory}'
 
         git config user.email "rootci@root.cern"
         git config user.name 'ROOT Continous Integration'
 
-        git fetch origin {head_ref_src}:{head_ref_dst}
-        git checkout {head_ref_dst}
+        git fetch {repository} {head_sha}:{head_ref}
+        git checkout {head_ref}
         git rebase {base_ref}
     """)
 
     if result != 0:
         die(result, "Rebase failed")
 
+# get_stdout_subprocess
+# execute and log a command.
+# capture the stdout, strip white space and return it
+# die in case of failed execution unless the error_message is empty.
+def get_stdout_subprocess(command: str, error_message: str) -> str:
+
+  result  = subprocess_with_capture(command)
+  if result.returncode != 0:
+    if error_message != "":
+      die(result, error_message)
+    else:
+      print("\033[90m", end='')
+      print(result.stdout)
+      print(result.stderr)
+      print("\033[0m", end='')
+      return ""
+  if result.stderr != "":
+    print("\033[90m", end='')
+    print(result.stdout)
+    print(result.stderr)
+    print("\033[0m", end='')
+  string_result = result.stdout
+  string_result = string_result.strip()
+  return string_result
+
+
+# get_base_head_sha
+#
+# Given a pull request merge commit and the incoming commit return
+# the commit corresponding to the head of the branch we are merging into.
+@github_log_group("Rebase")
+def get_base_head_sha(directory: str, repository: str, merge_sha: str, head_sha: str) -> str:
+
+  command = f"""
+      cd '{WORKDIR}/{directory}'
+      git fetch {repository} {merge_sha}
+      """
+  result = subprocess_with_log(command)
+  if result != 0:
+      die("Failed to fetch {merge_sha} from {repository}")
+  command = f"""
+      cd '{WORKDIR}/{directory}'
+      git rev-list --parents -1 {merge_sha}
+      """
+  result = get_stdout_subprocess(command, "Failed to find the base branch head for this pull request")
+
+  for s in result.split(' '):
+    if (s != merge_sha and s != head_sha):
+      return s
+
+  return ""
+
+@github_log_group("Pull/clone roottest branch")
+# relatedrepo_GetClosestMatch(REPO_NAME <repo> ORIGIN_PREFIX <originp> UPSTREAM_PREFIX <upstreamp>
```

Comment:
function docstrings should be written just below the function signature with triple quotes